plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergCombinedPageSource.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.Closer;
+import io.trino.spi.Page;
+import io.trino.spi.connector.ConnectorPageSource;
+import io.trino.spi.connector.EmptyPageSource;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.Collections.emptyIterator;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Concatenates the Pages for all PageSources which were grouped together an Iceberg CombinedScanTask.
+ */
+public class IcebergCombinedPageSource
+        implements ConnectorPageSource
+{
+    private final List<ConnectorPageSource> delegatePageSources;
+    private final Iterator<ConnectorPageSource> pageSourceIterator;
+    private final Closer closer = Closer.create();
+
+    private ConnectorPageSource currentPageSource;
+    private long completedBytes;
+    private long readTimeNanos;
+
+    public IcebergCombinedPageSource(List<ConnectorPageSource> delegatePageSources)
+    {
+        if (delegatePageSources.isEmpty()) {
```

Comment:
for simplicity we may  `checkArgument(!delegatePageSources.isEmpty()` to avoid dead code